src/crepe/system/CollisionSystem.cpp

```diff
@@ -1,5 +1,393 @@
+			resolution.y = (delta.y > 0) ? -overlap_y : overlap_y;
+		}
+	}
+
+	return resolution;
+}
+
+void CollisionSystem::determine_collision_handler(CollisionInfo& info){
+	// Check rigidbody type for static
+	if(info.first_rigidbody.data.body_type != Rigidbody::BodyType::STATIC)
+	{
+		// If second body is static perform the static collision handler in this system
+		if(info.second_rigidbody.data.body_type == Rigidbody::BodyType::STATIC){
+			static_collision_handler(info);
+		}; 
+		// Call collision event for user
+		CollisionEvent data(info);
+		EventManager::get_instance().trigger_event<CollisionEvent>(data, info.first_collider.game_object_id);
+	}		
+}
+
+void CollisionSystem::static_collision_handler(CollisionInfo& info){
+	// Move object back using calculate move back value 
+	info.first_transform.position += info.resolution;
+
+	// If bounce is enabled mirror velocity
+	if(info.first_rigidbody.data.bounce) {
+		if(info.resolution_direction == Direction::BOTH)
+		{
+			info.first_rigidbody.data.linear_velocity.y = -info.first_rigidbody.data.linear_velocity.y * info.first_rigidbody.data.elastisity;
+			info.first_rigidbody.data.linear_velocity.x = -info.first_rigidbody.data.linear_velocity.x * info.first_rigidbody.data.elastisity;
+		}
+		else if(info.resolution_direction == Direction::Y_DIRECTION) {
+			info.first_rigidbody.data.linear_velocity.y = -info.first_rigidbody.data.linear_velocity.y * info.first_rigidbody.data.elastisity;
+		}
+		else if(info.resolution_direction == Direction::X_DIRECTION){
+			info.first_rigidbody.data.linear_velocity.x = -info.first_rigidbody.data.linear_velocity.x * info.first_rigidbody.data.elastisity;
+		}
+	}
+	// Stop movement if bounce is disabled
+	else {
+		info.first_rigidbody.data.linear_velocity = {0,0};
+	}
+}
+
+std::vector<std::pair<CollisionSystem::CollisionInternal,CollisionSystem::CollisionInternal>> CollisionSystem::check_collisions(std::vector<collider_variant> & colliders) {
+	
+	
+	// TODO:
+	// If no colliders skip
+	// Check if colliders has rigidbody if not skip
+
+	// TODO:
+	// If amount is higer than lets say 16 for now use quadtree otwerwise skip
+	// Quadtree code
+	// Quadtree is placed over the input vector
+
+	std::vector<std::pair<CollisionInternal,CollisionInternal>> collisions_ret;
+	for (size_t i = 0; i < colliders.size(); ++i) {
+    std::visit([&](auto& inner_collider_ref) {
+			if (!inner_collider_ref.get().active) return;
+			auto inner_components = get_active_transform_and_rigidbody(inner_collider_ref.get().game_object_id);
+			if (!inner_components) return;
+			for (size_t j = i + 1; j < colliders.size(); ++j) {
+				std::visit([&](auto& outer_collider_ref) { 
+					if (!outer_collider_ref.get().active) return;
+					if (inner_collider_ref.get().game_object_id == outer_collider_ref.get().game_object_id) return;
+					auto outer_components = get_active_transform_and_rigidbody(outer_collider_ref.get().game_object_id);
+					if (!outer_components) return;
```

Comment:
Ik vind het super lelijk dat deze `return` eigenlijk functioneert als een `continue` van de for loop waarin je nu `std::visit` aanroept. Zou deze code geschreven kunnen worden zonder gebruik van `std::visit`? Ik heb deze functie nog nooit eerder gezien, en als ik de (soortvan onduidelijke) documentatie goed begrijp zou deze functie alleen de variant casten door middel van een visitor design pattern ofzo?